internal/systemd/systemd.go

```diff
@@ -0,0 +1,174 @@
+		if err != nil {
+			return nil, fmt.Errorf("cannot establish user connection to systemd: %v", err)
+		}
+	}
+
+	return &Conn{
+		ctx:  ctx,
+		conn: conn,
+	}, nil
+}
+
+func (c *Conn) Close() {
+	c.conn.Close()
+}
+
+// Reload instructs systemd to scan for and reload unit files.
+func (c *Conn) Reload() error {
+	return c.conn.ReloadContext(c.ctx)
+}
+
+// EnableUnit enables the named unit. If activate is true, it also starts the
+// unit. If runtime is true, the unit is enabled for the runtime only (/run). If
+// false, it is enabled persistently (/etc).
+func (c *Conn) EnableUnit(name string, activate bool, runtime bool) error {
+	if _, _, err := c.conn.EnableUnitFilesContext(c.ctx, []string{name}, runtime, true); err != nil {
+		return fmt.Errorf("cannot enable unit %v: %v", name, err)
+	}
+
+	if activate {
+		if err := c.StartUnit(name, true); err != nil {
+			return fmt.Errorf("cannot start unit %v: %v", name, err)
+		}
+	}
+
+	return nil
+}
+
+// StartUnit starts the named unit. If wait is true, the method waits until the
+// unit state becomes "active".
+func (c *Conn) StartUnit(name string, wait bool) error {
+	jobComplete := make(chan string)
+	_, err := c.conn.StartUnitContext(c.ctx, name, "replace", jobComplete)
+	if err != nil {
+		return fmt.Errorf("cannot start unit %v: %v", name, err)
+	}
+	result := <-jobComplete
+	switch result {
+	case "done":
+		// The job successfully started, break to proceed
+		break
+	default:
+		return fmt.Errorf("failed to start unit with reason: %v", result)
+	}
+
+	if wait {
+		if err := c.waitForState(name, "active", 1*time.Second); err != nil {
+			return fmt.Errorf("timed out waiting for state 'active': %v", err)
+		}
+	}
+
+	return nil
+}
+
+// DisableUnit disables the named unit. If deactivate is true, it also stops the
+// unit. If runtime is true, the unit is enabled for the runtime only (/run). If
+// false, it is enabled persistently (/etc).
+func (c *Conn) DisableUnit(name string, deactivate bool, runtime bool) error {
+	if _, err := c.conn.DisableUnitFilesContext(c.ctx, []string{name}, runtime); err != nil {
+		return fmt.Errorf("cannot disable unit %v: %v", name, err)
+	}
+
+	if deactivate {
+		if err := c.StopUnit(name, true); err != nil {
+			return fmt.Errorf("cannot stop unit %v: %v", name, err)
+		}
+	}
+
+	return nil
+}
+
+// StopUnit stops the named unit. If wait is true, the method waits until the
+// unit state becomes "inactive".
+func (c *Conn) StopUnit(name string, wait bool) error {
+	jobComplete := make(chan string)
+	_, err := c.conn.StopUnitContext(c.ctx, name, "replace", jobComplete)
+	if err != nil {
+		return fmt.Errorf("cannot stop unit %v: %v", name, err)
+	}
+	result := <-jobComplete
+	switch result {
+	case "done":
+		// The job successfully started, break to proceed
+		break
+	default:
+		return fmt.Errorf("failed to stop unit with reason: %v", result)
+	}
+
+	if wait {
+		if err := c.waitForState(name, "inactive", 1*time.Second); err != nil {
```

Comment:
Should we allow for this timeout value to be user configured? If it is timing out for some unknown reason for a given customer it might be valuable to provide a mechanism that allow them to lengthen the timeout length.

Thoughts?

Reply:
I thought about this when I wrote it. It does make sense to make it
customizable, but I struggled with trying to come up with a scenario where it
really would matter. This isn't a timeout waiting for a network call to resolve;
this is a local D-Bus RPC call. Those messages themselves have a timeout of 1
second by default too I think. The system would have to be under extremely
heavy load for local RPC socket traffic to be under resource constraints.

Would it make more sense to make it a longer default? Like 30 seconds?

Reply:
30 seconds may be too large. If the 1 second is the default for the underlying code perhaps 3 of 5 seconds will be sufficient without being too long.
Thoughts?

Reply:
I set the value to `5*time.Second`.